lib/puppet/pops/types/p_object_type.rb

```diff
@@ -0,0 +1,539 @@
+  #
+  # @return [Hash{String=>Object}] the features hash
+  # @api public
+  def i12n_hash
+    result = {}
+    result[KEY_PARENT] = @parent unless @parent.nil?
+    result[KEY_ATTRIBUTES] = compressed_members_hash(@attributes) unless @attributes.empty?
+    result[KEY_FUNCTIONS] = compressed_members_hash(@functions) unless @functions.empty?
+    result[KEY_EQUALITY] = @equality unless @equality.nil?
+    result[KEY_CHECKS] = @checks unless @checks.nil?
+    result[KEY_ANNOTATIONS] = @annotations unless @annotations.nil?
+    result
+  end
+
+  def eql?(o)
+    self.class == o.class &&
+      @parent == o.parent &&
+      @attributes == o.attributes &&
+      @functions == o.functions &&
+      @equality == o.equality &&
+      @checks == o.checks &&
+      @annotations == o.annotations
+  end
+
+  def hash
+    @hash
+  end
+
+  def kind_of_callable?(optional=true, guard = nil)
+    @parent.nil? ? false : @parent.kind_of_callable?(optional, guard)
+  end
+
+  def instance?(o, guard = nil)
+    assignable(TypeCalculator.infer(o), guard)
+  end
+
+  def iterable?(guard = nil)
+    @parent.nil? ? false : @parent.iterable?(guard)
+  end
+
+  def iterable_type(guard = nil)
+    @parent.nil? ? false : @parent.iterable_type(guard)
+  end
+
+  # Returns the members (attributes and functions) of this `Object` type. If _include_parent_ is `true`, then all
+  # inherited members will be included in the returned `Hash`.
+  #
+  # @param include_parent [Boolean] `true` if inherited members should be included
+  # @return [Hash{String=>PAnnotatedMember}] a hash with the members
+  # @api public
+  def members(include_parent = false)
+    get_members(include_parent, :both)
+  end
+
+  # Returns the attributes of this `Object` type. If _include_parent_ is `true`, then all
+  # inherited attributes will be included in the returned `Hash`.
+  #
+  # @param include_parent [Boolean] `true` if inherited attributes should be included
+  # @return [Hash{String=>PAttribute}] a hash with the attributes
+  # @api public
+  def attributes(include_parent = false)
+    get_members(include_parent, :attributes)
+  end
+
+  # Returns the attributes that participate in equality comparison. Inherited equality attributes
+  # are included.
+  # @return [Hash{String=>PAttribute}] a hash of attributes
+  # @api public
+  def equality_attributes
+    all = {}
+    collect_equality_attributes(all)
+    all
+  end
+
+  # @return [Boolean] `true` if this type is included when comparing instances
+  # @api public
+  def equality_include_type?
+    @equality_include_type
+  end
+
+  # Returns the functions of this `Object` type. If _include_parent_ is `true`, then all
+  # inherited functions will be included in the returned `Hash`.
+  #
+  # @param include_parent [Boolean] `true` if inherited functions should be included
+  # @return [Hash{String=>PFunction}] a hash with the functions
+  # @api public
+  def functions(include_parent = false)
```

Comment:
There is a #members method that returns both attributes and functions (line 417). Mixing that with other features (equality, annotations, checks) is IMO pointless since they don't live in the same namespace.